[Bugfix:Submission] Fix viewing/using older notebook versions

loadedskiffer opened the pull request:
### What is the current behavior?


Currently if a Student is working on a Notebook gradeable and they choose to view an older submission, their most recent answers are displayed instead of the ones that they used for the version they chose to view. 

When a student gets a test back, if they specified an older version as "GRADE THIS VERSION"  the answers they gave for the most recent  submission will be displayed instead of the ones from the selected version. At the bottom of the page in the "Select Submission Version" box, the most recent submission will be shown as selected, with the "GRADE THIS VERSION" .

### What is the new behavior?
closes #6080 
closes #5288
closes #6306

The answers a student gave for a specific submission version will be correctly displayed when choosing between viewing different versions.
 All buttons (clear, reset, recent versions ) and inputs (radio, multiple select, short answer, codebox) are disabled when a student is viewing an older version. 
<img width="289" alt="pr11_1" src="https://user-images.githubusercontent.com/66340271/123852002-d48e3080-d8e9-11eb-93cf-a8187c592638.PNG">

If there is any provided code for a short answer/codebox the "clear" button now says "reset" to eliminate confusion.
<img width="537" alt="pr11_2" src="https://user-images.githubusercontent.com/66340271/123852086-ebcd1e00-d8e9-11eb-998f-51b379547865.PNG">

When a student gets a test back, if they specified an older version as "GRADE THIS VERSION"  the answers they gave for that submission will be displayed correctly. At the bottom of the page in the "Select Submission Version" box, the correct submission will be shown as selected, with the "GRADE THIS VERSION" .

The "submit" button now says "Switch To Most Recent Version To Make New Submission" when viewing an older submission.
### Other information?
Testing is easiest with "Notebook basic" in the development course because it has every type of question.

williamjallen commented on site/app/models/notebook/Notebook.php:
```diff
@@ -155,28 +155,37 @@ private function getMarkdownData($cell) {
+                            //get answer for the selected display version
+                            $question_name = $notebookVal['filename'];
+                            $file = fopen("/var/local/submitty/courses/" . $semester . "/" . $course . "/submissions/" . $gradeable_id . "/" . $student_id . "/" . $display_version . "/" . $question_name, "r");
+                            $version_answer = fread($file, filesize("/var/local/submitty/courses/" . $semester . "/" . $course . "/submissions/" . $gradeable_id . "/" . $student_id . "/" . $display_version . "/" . $question_name));
```
It would be better to use `$this->core->getConfig()->getCoursePath()` here instead of hardcoded paths.  That would also mean that you wouldn't need the `$semester` and `$course` parameters I believe.

williamjallen commented on site/app/models/notebook/Notebook.php:
```diff
@@ -187,9 +196,15 @@ public function getMostRecentNotebookSubmissions(int $version, array $new_notebo
+                            //get answer for the selected display version
+                            $question_name = $notebookVal['filename'];
+                            $file = fopen("/var/local/submitty/courses/" . $semester . "/" . $course . "/submissions/" . $gradeable_id . "/" . $student_id . "/" . $display_version . "/" . $question_name, "r");
+                            $version_answer = fread($file, filesize("/var/local/submitty/courses/" . $semester . "/" . $course . "/submissions/" . $gradeable_id . "/" . $student_id . "/" . $display_version . "/" . $question_name));
```
Same thing here regarding the paths

williamjallen commented on site/app/views/grading/ElectronicGraderView.php:
```diff
@@ -973,7 +973,11 @@ public function hwGradingPage(Gradeable $gradeable, GradedGradeable $graded_grad
+                $display_version,
+                $graded_gradeable->getGradeableId(),
+                $this->core->getConfig()->getSemester(),
+                $this->core->getConfig()->getCourse()
```
These would be able to be removed if the paths were changed.

eddiekrystowski reviewed (changes requested):
Mostly tried to break the UI on my review. I was able to break it by following these steps to get around the disabled inputs:
1. Open console
2. change the value of any input using Javascript/JQuery/editing the HTML. For example, I chose a different option on a multiple choice input with `$('#multiple_choice_0_1').prop('checked', 'true');`
3. set `empty_inputs = false;`
4. Manually trigger form submission with `$('#submit').trigger('click')`
5. The page now reloads to the newly made version with some php errors, and some of the inputs (in my case radio/checkbox inputs) have been emptied.

I'm not sure what the intended behavior is here, but perhaps there is a way to just disallow any submission going through on the backend if they are on an inactive version. This also doesn't break the gradeable, they can still scroll past the php errors and resubmit again and the gradeable will be fine.

williamjallen commented on site/app/models/notebook/Notebook.php:
```diff
@@ -171,8 +173,10 @@ public function getMostRecentNotebookSubmissions(int $version, array $new_notebo
-                            $file = fopen("/var/local/submitty/courses/" . $semester . "/" . $course . "/submissions/" . $gradeable_id . "/" . $student_id . "/" . $display_version . "/" . $question_name, "r");
-                            $version_answer = fread($file, filesize("/var/local/submitty/courses/" . $semester . "/" . $course . "/submissions/" . $gradeable_id . "/" . $student_id . "/" . $display_version . "/" . $question_name));
+                            $file_path = $this->core->getConfig()->getCoursePath();
+                            $file_path = $file_path . "/submissions/" . $gradeable_id . "/" . $student_id . "/" . $display_version . "/" . $question_name;
```
I overlooked this in my last review but you could also use `FileUtils::joinPaths()` for these paths instead of manually concatenating it.

williamjallen commented on site/app/models/notebook/Notebook.php:
```diff
@@ -198,8 +202,10 @@ public function getMostRecentNotebookSubmissions(int $version, array $new_notebo
-                            $file = fopen("/var/local/submitty/courses/" . $semester . "/" . $course . "/submissions/" . $gradeable_id . "/" . $student_id . "/" . $display_version . "/" . $question_name, "r");
-                            $version_answer = fread($file, filesize("/var/local/submitty/courses/" . $semester . "/" . $course . "/submissions/" . $gradeable_id . "/" . $student_id . "/" . $display_version . "/" . $question_name));
+                            $file_path = $this->core->getConfig()->getCoursePath();
+                            $file_path = $file_path . "/submissions/" . $gradeable_id . "/" . $student_id . "/" . $display_version . "/" . $question_name;
```
same here

eddiekrystowski reviewed (approved):
Works well, I can't break it with the same method as before.

MasterOdin commented on site/app/models/notebook/Notebook.php:
```diff
@@ -154,29 +154,42 @@ private function getMarkdownData($cell) {
+    * @param string $gradeable_id
     */
-    public function getMostRecentNotebookSubmissions(int $version, array $new_notebook, string $student_id): array {
+    public function getMostRecentNotebookSubmissions(int $version, array $new_notebook, string $student_id, int $display_version, string $gradeable_id): array {
```
You've typed this as `int` here but pass in `null` below? If `null` is a valid value, you need to add a `?` before the type, like `?int`.

MasterOdin commented on site/app/views/submission/HomeworkView.php:
```diff
@@ -315,13 +315,29 @@ private function renderSubmitBox(Gradeable $gradeable, $graded_gradeable, $versi
+                        $graded_gradeable->getAutoGradedGradeable()->getHighestVersion(),
+                        $notebook,
+                        $this->core->getUser()->getId(),
+                        $version_instance->getVersion() ?? null,
```
Can `getVersion()` return something other than `?int` that we need the null coaelescing operator here?

MasterOdin replied:
Ignoring the comment above, why have the null coalescing operator here? What is it doing?

loadedskiffer replied:
looks like it's doing nothing, I didn't write that line so I didn't look into it before.

MasterOdin commented on site/app/views/submission/HomeworkView.php:
```diff
@@ -417,6 +432,11 @@ private function renderSubmitBox(Gradeable $gradeable, $graded_gradeable, $versi
 
         $highest_version = $graded_gradeable !== null ? $graded_gradeable->getAutoGradedGradeable()->getHighestVersion() : 0;
 
+        $viewing_inactive_version = false;
+        if ($highest_version !== $display_version) {
+            $viewing_inactive_version = true;
+        }
```
```suggestion
        $viewing_inactive_version = $highest_version !== $display_version;
```

MasterOdin commented on site/public/js/gradeable-notebook.js:
```diff
@@ -194,15 +190,26 @@ $(document).ready(() => {
         }
         else {
-            $(clear_button_id).attr('disabled', false);
+            const value = $(clear_button_id).attr('data-older_version');
+            if (value) {
+                $(clear_button_id).attr('disabled', true);
+            }
+            else {
+                $(clear_button_id).attr('disabled', false);
+            }
```
Can condense by coercing `value` to a boolean using `!!` notation:
```suggestion
            $(clear_button_id).attr('disabled', !!$(clear_button_id).attr('data-older_version'));
```

MasterOdin commented on site/public/js/gradeable-notebook.js:
```diff
@@ -194,15 +190,26 @@ $(document).ready(() => {
         else {
-            $(recent_button_id).attr('disabled', false);
-            $('#submit').attr('disabled', false);
+            const value = $(recent_button_id).attr('data-older_version');
+            if (value) {
+                $(recent_button_id).attr('disabled', true);
+            }
+            else {
+                $(recent_button_id).attr('disabled', false);
+            }
```
same here

MasterOdin commented on site/app/templates/submission/homework/SubmitBox.twig:
```diff
@@ -341,8 +342,11 @@
         </div>
-
-        <button type="button" id="submit" class="btn btn-success">Submit</button>
+        {% if viewing_inactive_version %}
+            <button type="button" id="submit" class="btn btn-success">Switch To Most Recent Version To Make New Submission</button>
+        {% else %}
+            <button type="button" id="submit" class="btn btn-success">Submit</button>
+        {% endif %}
```
can condense

MasterOdin commented on site/app/templates/notebook/Notebook.twig:
```diff
@@ -230,11 +277,18 @@
-                    Clear
-                </button>
-
+                {% if viewing_inactive_version %}
+                    <button type="button" class="btn btn-primary
+                        sa-clear-reset" onclick="deleteFiles({{num_file_submissions}});"
+                        disabled = "disabled">
+                        Clear
+                    </button>
+                {% else %}
+                    <button type="button" class="btn btn-primary
+                        sa-clear-reset" onclick="deleteFiles({{num_file_submissions}});">
+                        Clear
+                    </button>
+                {% endif %}
```
can condense

MasterOdin commented on site/app/templates/notebook/Notebook.twig:
```diff
@@ -177,22 +214,32 @@
                     {% else %}
-                        $("#mc_{{ num_multiple_choice }}_clear_button").attr("disabled", false);
-                        $("#mc_{{ num_multiple_choice }}_recent_button").attr("disabled", true);
+                        {% if viewing_inactive_version %}
+                            $("#mc_{{ num_multiple_choice }}_clear_button").attr("disabled", true);
+                            $("#mc_{{ num_multiple_choice }}_recent_button").attr("disabled", true);
+                        {% else %}
+                            $("#mc_{{ num_multiple_choice }}_clear_button").attr("disabled", false);
+                            $("#mc_{{ num_multiple_choice }}_recent_button").attr("disabled", true);
+                        {% endif %}
```
condense

MasterOdin commented on site/app/templates/notebook/Notebook.twig:
```diff
@@ -177,22 +214,32 @@
-                <button type="button" id="mc_{{ num_multiple_choice }}_clear_button" class="btn btn-primary mc-clear">Clear</button>
-                <button type="button" id="mc_{{ num_multiple_choice }}_recent_button" class="btn btn-primary mc-recent">Use Most Recent Submission</button>
-
+                {% if viewing_inactive_version %}
+                    <button type="button" id="mc_{{ num_multiple_choice }}_clear_button" class="btn btn-primary mc-clear" disabled="disabled">Clear</button>
+                    <button type="button" id="mc_{{ num_multiple_choice }}_recent_button" class="btn btn-primary mc-recent" disabled="disabled">Use Most Recent Submission</button>
+                {% else %}
+                    <button type="button" id="mc_{{ num_multiple_choice }}_clear_button" class="btn btn-primary mc-clear">Clear</button>
+                    <button type="button" id="mc_{{ num_multiple_choice }}_recent_button" class="btn btn-primary mc-recent">Use Most Recent Submission</button>
+                {% endif %}
```
condense

MasterOdin commented on site/app/templates/notebook/Notebook.twig:
```diff
@@ -134,17 +164,24 @@
-                                    "content" : cell.choices[idx].description
-                                } only %}
+                            <label for="multiple_choice_{{ num_multiple_choice }}_{{ loop.index0 }}" class="markdown">
+                                {% if viewing_inactive_version %}
+                                    <input type="checkbox" class="key_to_click" tabindex="0"
+                                           name="multiple_choice_{{ num_multiple_choice }}"
+                                           id="multiple_choice_{{ num_multiple_choice }}_{{ loop.index0 }}"
+                                           value="{{ cell.choices[idx].value }}"
+                                           onclick="handle_input_keypress({{ viewing_inactive_version }});"
+                                           disabled = "disabled"
+                                    />
+                                {% else %}
+                                    <input type="checkbox" class="key_to_click" tabindex="0"
+                                           name="multiple_choice_{{ num_multiple_choice }}"
+                                           id="multiple_choice_{{ num_multiple_choice }}_{{ loop.index0 }}"
+                                           value="{{ cell.choices[idx].value }}"
+                                           onclick="handle_input_keypress({{ viewing_inactive_version }});"
+                                    />
+                                {% endif %}
```
condense

MasterOdin commented:
@loadedskiffer looks like there's still a number of unaddressed comments above. Looks like they're hidden by default and you'll need to hit the "show more" action to view them.

loadedskiffer commented:
> @loadedskiffer looks like there's still a number of unaddressed comments above. Looks like they're hidden by default and you'll need to hit the "show more" action to view them.

sorry about that, I didn't see them before.
